testplan/web_ui/testing/src/Report/BatchReportBeta/components/NavSidebarWithNextRoute.jsx

```diff
@@ -0,0 +1,134 @@
+    // Then the value of the following variables are:
+    // *       url = "/aaa/bbb/ccc/12345"
+    // *      path = "/aaa/bbb/ccc/:id"
+    // * params.id = "12345"
+    //
+    // Note that `match === null` if the current component was rendered like:
+    //
+    //   <Route exact path={"/a/b/c"} children={ThisComponent} />
+    //
+    // ... and we're not at "/a/b/c" because the `children` prop always renders
+    // even when there's not match
+    let tgtEntry;
+    return {
+      passedLast, isLast, abortRender,
+      routePath: matchIsObj ? match.url : null,
+      getTgtEntryEntries: () => {
+        if(abortRender) return null;
+        if(_isUndefined(tgtEntry)) {
+          if(entriesIsObj) {
+            if(_isArray(entries)) {
+              const encodedName = match.params.id;
+              // this incoming `encodedName` may be URL-encoded and so it
+              // won't match `entry.name` in the `entries` array, so we grab
+              // whatever `id` is actually an alias for, and use that to find
+              // our target `entry` object.
+              let decodedName = hashComponentToAlias[encodedName];
+              // on refresh on an aliased path, the `componentAliases` will
+              // be empty so we need to fill it with the aliased component
+              if(!decodedName) {
+                decodedName = uriComponentCodec.decode(encodedName);
+                boundSetHashComponentAlias(encodedName, decodedName);
+              }
+              tgtEntry = entries.find(e => decodedName === e.name);
+            } else {
+              tgtEntry = entries;
+            }
+          } else {
+            tgtEntry = null;
+          }
+        }
+        return _isObject(tgtEntry) ? tgtEntry.entries : null;
+      },
+    };
+  },
+);
+
+const NavSidebarWithNextRoute = connector(
+  ({ getTgtEntryEntries, routePath, passedLast, isLast, abortRender }) => (
+    abortRender ? null : (
+      /**
+       * To illustrate how this works, say we have these possible routes:
+       * 1. /
+       * 2. /a
+       * 3. /a/b
+       * 4. /a/b/c
+       * 5. /a/b/d
+       *
+       * ... TBC
+       */
+
+        <>
+          {
+            /**
+             * ... TBC
+             */
+          }
+          <Route exact
+                 path={routePath}
+                 render={({ location, match }) => (
+                   <NavSidebar entries={getTgtEntryEntries()}
+                               location={location}
+                               match={match}
+                   />
+                 )}
+          />
+          {
+            /**
+             * ... TBC
+             *
+             * Note that, the last time we pass through the below Route, we will
+             * be one-passed the end of the entries that can go into the sidebar,
+             * i.e. the next time BoundNavSidebarWithNextRoute renders, it will
+             * receive entries that
+             */
+          }
```

Comment:
We have 2 recursive router in Breadcrumbs and NavSidebar. Can we possibly just have 1 router at higher level to do the routing? https://ui.dev/react-router-v4-sidebar-breadcrumbs/

Reply:
1st priority for me after this gets in-product is to merge NavPanes.jsx, NavSidebar.jsx, NavSidebarWithNextRoute.jsx, NavBreadcrumb.jsx, and NavBreadcrumbWithNextRoute.jsx into one or two uncomplicated components.

Believe it or not it took a lot of trial and error  to get to the complicated mess you see before you. The article is interesting but the toughest part is keeping the breadcrumbs and sidebar reliably "in-sync-but-off-by-one".

React Router v6 was released about a month ago and its new relative nested routes feature should uncomplicate the routing throughout the app by a lot.